<img src="skk_gsb1.png"></img>

<center><font size="6">Team Project 1 - Content vs Price Sensitive Customer</font></center>
<p>
<center><font size="3">Business Forecast by Yanbo Wang</font></center>

<center><font size="3">Team 6: Park Keuntae, Nur Afiqah binti Ahmad Shukri</font></center>

# Background

## Business Context

You are working for an **online book platform** that offers both paid and free content. The company’s leadership wants to **differentiate** customers who are **content/experience-driven** (i.e., they buy books and engage deeply with them) from those who are **price-driven** (i.e., they focus on free content, sampling, or finding deals).

**Business Problem**: By identifying whether a customer is content- or price-driven, the marketing team can tailor promotions, upsell strategies, and new feature developments.

1. **Content/experience-driven customers** might respond better to:
   - Exclusive previews or early access to premium books.  
   - Advanced reading features (e.g., curated reading lists, advanced highlighting capabilities).

2. **Price-driven customers** might respond better to:
   - Discount offers or bundle deals.  
   - Extended free trials or “limited-time” free chapters.

---

## Why This Matters

- **Revenue Maximization**: If you misclassify a content-driven customer as price-driven, you might offer unneeded discounts and lose revenue.  
- **Customer Satisfaction**: Price-driven customers pushed with expensive premium items may feel alienated. Conversely, content-driven customers might be willing to pay more for better reading experiences—and you want to serve them accordingly.  
- **Long-term Loyalty**: Matching the right offers to the right user segment helps increase retention (reduce churn) and fosters loyalty.

---



# Dataset

## Data Files

- team_project_1.csv : This file contains data for past customers who used liveBook and whether or not they indicate in a survay that they care content more or price more. For each customer, there is a set of measurements about the use of liveBook by the customers. These measurements are described in more detail below.

## Variables in Data

The measurements of the customers, also known as the “features” of the dataset, are described below:

- Content Driven Customer: 1 if they indicate in a survay that they care content more, 0 if care price more.

- readingownedbook: This is how many events of reading a book the customer-owned.

- ebookdownloaded: This is how many times the customer downloads an e-book.

- readingfreepreview : This is how many times the customer read a free preview.

- highlightcreated: This is how many highlights the customer created.

- freecontentcheckout: This is how many times the customer used the free content check-out feature.

- readingopenchapter: This is how many times the customer read an open (free) chapter.

- wishlistitemadded: This is how many times the customer added items to their wish list.

- crossreferencetermopened: This is how many times the customer opened a cross-reference term.

- total_events: This is the total number of events logged for the customer.

- distinct_products: This is how many distinct products the customer purchased in the previous period.

- reading_utilities: This is how many times the customer used all reading utilities, including highlights and cross-reference terms.

- downloads_per_book: This is how many e-book downloads the customer had divided by the number of books they owned.

- percent_reading: This is the percent of the customers’ events that included reading.

- percent_utilities: This is the percent of the customers’ events that were utility features.

- events_per_book: This is the number of events the customer had divided by the distinct number of books they purchased.


## Data Collection Process

We recruited a select group of existing customers and offered them a modest incentive (e.g., gift cards, account credits) to complete a short survey about their primary motivation when purchasing books on our platform. The survey asked targeted questions such as:

* How important is reading quality, features (like highlighting or cross-referencing), and overall experience in your buying decision?
* How important are discounts, promotions, and overall price when making a purchase?
* Based on their answers, we categorized each respondent as either content/experience-driven or price-driven. This survey-labeled dataset now serves as a valuable “ground truth” for building and validating our predictive models on broader customer data.

# Team Project Tasks



## Task A: Regression Without Interaction Terms

1. **Set Up a Linear Regression**: Use a package like `statsmodels` to create a simple regression model.  
   - **Features**: Use your set of features *excluding* newly engineered interactions.  
   - **Target**: `y` (the dummy variable: 1 if content-driven, 0 if price-driven).  
2. **Fit the Model** and observe:
   - The coefficient estimates.  
   - The significance of each coefficient (if using `statsmodels`).  

**Question**: Which features appear most predictive of a user being content-driven vs. price-driven?

---

## Task B: Regression With Interaction Terms

1. **Enhance the Model**: Now **add** the interaction terms you created in Task B to your feature set.  
2. **Fit a New Model**. Compare:
   - Changes in the R-squared (or other evaluation metrics).  
   - Changes in the significance and magnitude of the main effects when interaction terms are introduced.  
   - Which interactions turned out to be strong predictors?

---

## Task C: Interpretation and Recommendations

1. **Which Features Matter Most?** Summarize the regression results:
   - Positive drivers (can be individual features or interaction of features) of a user being content-driven.  
   - Negative drivers (features that indicate price sensitivity).  
2. **What Are the Practical Implications?** Suppose you’re presenting to the marketing team:
   - How might you **segment customers** or **tailor promotions** based on these findings?  
   - Which features or behaviors would you track more closely going forward?

**Hint**:

**It is possible that one feature might affect the predicted value alone.** For example:

High readingownedbook: These customers spend a lot of time reading books they have already purchased, indicating they deeply value the reading experience.

**It is equally possible that two features jointly affect the predicted value (i.e. interaction effects of features).** For example,

A strong interaction may occur between percent_reading and highlightcreated or crossreferencetermopened. A high proportion of reading events (percent_reading) plus active use of reading utilities is an even stronger indicator of someone who’s driven by the reading experience.

**Deliverable**: A 2-3 page **report** explaining:
- **Key takeaways** from the regression analysis.  
- **Practical suggestions** on how to leverage these insights in marketing, product development, or customer support.

**Good luck**, and we look forward to reviewing your insights on how to identify and serve these two distinct customer segments!

# Predict Customer Churn with OLS Linear Regression

## Task A: Regression Analysis without Interaction Effects

We use the first 20000 rows to fit regression model. We hold back the remaining rows for the subsequent model evaluation purpose.

In [1]:
features = [
    'readingownedbook', 'ebookdownloaded', 'readingfreepreview',
    'highlightcreated', 'freecontentcheckout', 'readingopenchapter',
    'wishlistitemadded', 'crossreferencetermopened', 'total_events',
    'distinct_products', 'reading_utilities', 'downloads_per_book',
    'percent_reading', 'percent_utilities', 'events_per_book'
]

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
import pandas as pd

df=pd.read_csv('/content/drive/MyDrive/BF/team_project_1.csv')
df=df.iloc[:20000, :]
df.head()

,account_id,metric_time,readingownedbook,ebookdownloaded,readingfreepreview,highlightcreated,freecontentcheckout,readingopenchapter,wishlistitemadded,crossreferencetermopened,total_events,distinct_products,reading_utilities,downloads_per_book,percent_reading,percent_utilities,events_per_book,Content Driven Customer
0,0001b5d2fd33d5b98539d39658d69a90,2020-03-01,13.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,16.0,2.0,0.0,0.0,0.8125,0.0,8.0,1
1,00036f182bb0746fc8048d8e0bc785b3,2020-03-01,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0000,0.0,1.0,0
2,0006aabe0ba47a35c0b0bf6596f85159,2020-03-01,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0000,0.0,1.0,0
3,0006d94927acb83fdfae843f80d54507,2020-03-01,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,2.0,1.0,0.0,1.0,0.0000,0.0,2.0,1
4,000aa02f54c61c18c94096fc563115f8,2020-03-01,0.0,4.0,0.0,0.0,0.0,0.0,0.0,0.0,4.0,2.0,0.0,2.0,0.0000,0.0,2.0,0


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 18 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   account_id                20000 non-null  object 
 1   metric_time               20000 non-null  object 
 2   readingownedbook          20000 non-null  float64
 3   ebookdownloaded           20000 non-null  float64
 4   readingfreepreview        20000 non-null  float64
 5   highlightcreated          20000 non-null  float64
 6   freecontentcheckout       20000 non-null  float64
 7   readingopenchapter        20000 non-null  float64
 8   wishlistitemadded         20000 non-null  float64
 9   crossreferencetermopened  20000 non-null  float64
 10  total_events              20000 non-null  float64
 11  distinct_products         20000 non-null  float64
 12  reading_utilities         20000 non-null  float64
 13  downloads_per_book        20000 non-null  float64
 14  percen

In [7]:
import statsmodels.api as sm
import numpy as np

# Define X and y
X = df[features]
y = df['Content Driven Customer']

# Add a constant to the model(Adding y-intercept)
X = sm.add_constant(X)

# Fit the OLS model
model = sm.OLS(y, X).fit()
print(model.summary())

# Summary table into a DataFrame
summary_df = pd.DataFrame({
    "Variable": model.params.index,
    "Coef": model.params.values,
    "Std Err": model.bse.values,
    "t": model.tvalues,
    "P>|t|": model.pvalues,
    "CI 2.5%": model.conf_int()[0],
    "CI 97.5%": model.conf_int()[1]
})

                               OLS Regression Results                              
Dep. Variable:     Content Driven Customer   R-squared:                       0.305
Model:                                 OLS   Adj. R-squared:                  0.305
Method:                      Least Squares   F-statistic:                     627.3
Date:                     Thu, 13 Feb 2025   Prob (F-statistic):               0.00
Time:                             06:26:46   Log-Likelihood:                -10789.
No. Observations:                    20000   AIC:                         2.161e+04
Df Residuals:                        19985   BIC:                         2.173e+04
Df Model:                               14                                         
Covariance Type:                 nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------

## Regression Analysis with Interactions Effects

## **Question: Which features appear most predictive of a user being content-driven vs. price-driven?**

Based on the regression output, the key predictors determine if customers are content-driven (1) or price-driven (2). Below are regression output of customers being content-driven.


(1) Content-Sensitive Indicators:

>*Percent_reading* is the most predictive feature for determining whether a user is content-driven or price-driven.
>
>- percent_reading (Coefficient:0.6803, p-value: 0.0000), indicates a strong interest into reading are more likely to be content-driven.
>
>Other significant features that customers are being content-driven include:
>- percent_utilities (Coefficient: 0.5469, p-value: 0.0000), indicating strong association with utility feature usage.
- distinct_products (Coefficient: 0.0232, p-value: 0.0000), showing a positive correlation with purchasing diverse products.

(2) Price-Sensitive Indicators:

>- freecontentcheckout (Coefficient: -0.0113, p-value: 0.000), users who frequently check out free content are less likely to be content-driven, reinforcing price sensitivity.




# Task B : Regression with Interaction Terms

We use the first 20000 rows to fit regression model. We hold back the remaining rows for the subsequent model evaluation purpose.

In [8]:
features = [
    'readingownedbook', 'ebookdownloaded', 'readingfreepreview',
    'highlightcreated', 'freecontentcheckout', 'readingopenchapter',
    'wishlistitemadded', 'crossreferencetermopened', 'total_events',
    'distinct_products', 'reading_utilities', 'downloads_per_book',
    'percent_reading', 'percent_utilities', 'events_per_book'
]

In [10]:
import pandas as pd

df=pd.read_csv('/content/drive/MyDrive/BF/team_project_1.csv')
df=df.iloc[:20000, :]
df.head()

,account_id,metric_time,readingownedbook,ebookdownloaded,readingfreepreview,highlightcreated,freecontentcheckout,readingopenchapter,wishlistitemadded,crossreferencetermopened,total_events,distinct_products,reading_utilities,downloads_per_book,percent_reading,percent_utilities,events_per_book,Content Driven Customer
0,0001b5d2fd33d5b98539d39658d69a90,2020-03-01,13.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,16.0,2.0,0.0,0.0,0.8125,0.0,8.0,1
1,00036f182bb0746fc8048d8e0bc785b3,2020-03-01,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0000,0.0,1.0,0
2,0006aabe0ba47a35c0b0bf6596f85159,2020-03-01,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0000,0.0,1.0,0
3,0006d94927acb83fdfae843f80d54507,2020-03-01,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,2.0,1.0,0.0,1.0,0.0000,0.0,2.0,1
4,000aa02f54c61c18c94096fc563115f8,2020-03-01,0.0,4.0,0.0,0.0,0.0,0.0,0.0,0.0,4.0,2.0,0.0,2.0,0.0000,0.0,2.0,0


In [11]:
import pandas as pd
import itertools
import statsmodels.api as sm
import numpy as np

# Generate all possible pairs of features
feature_pairs = list(itertools.combinations(features, 2))

interaction_features = []
# Create interaction terms for each pair
for feature1, feature2 in feature_pairs:
    interaction_name = f"{feature1} x {feature2}"
    interaction_features.append(interaction_name)
    df[interaction_name] = df[feature1] * df[feature2]

# Display the first few rows of the DataFrame with the new interaction columns
df.head()

<ipython-input-11-1613f792f1c9>:14: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df[interaction_name] = df[feature1] * df[feature2]
<ipython-input-11-1613f792f1c9>:14: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df[interaction_name] = df[feature1] * df[feature2]
<ipython-input-11-1613f792f1c9>:14: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fra

,account_id,metric_time,readingownedbook,ebookdownloaded,readingfreepreview,highlightcreated,freecontentcheckout,readingopenchapter,wishlistitemadded,crossreferencetermopened,...,reading_utilities x downloads_per_book,reading_utilities x percent_reading,reading_utilities x percent_utilities,reading_utilities x events_per_book,downloads_per_book x percent_reading,downloads_per_book x percent_utilities,downloads_per_book x events_per_book,percent_reading x percent_utilities,percent_reading x events_per_book,percent_utilities x events_per_book
0,0001b5d2fd33d5b98539d39658d69a90,2020-03-01,13.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,6.5,0.0
1,00036f182bb0746fc8048d8e0bc785b3,2020-03-01,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0
2,0006aabe0ba47a35c0b0bf6596f85159,2020-03-01,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0
3,0006d94927acb83fdfae843f80d54507,2020-03-01,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,2.0,0.0,0.0,0.0
4,000aa02f54c61c18c94096fc563115f8,2020-03-01,0.0,4.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,4.0,0.0,0.0,0.0


In [12]:
print(interaction_features)

['readingownedbook x ebookdownloaded', 'readingownedbook x readingfreepreview', 'readingownedbook x highlightcreated', 'readingownedbook x freecontentcheckout', 'readingownedbook x readingopenchapter', 'readingownedbook x wishlistitemadded', 'readingownedbook x crossreferencetermopened', 'readingownedbook x total_events', 'readingownedbook x distinct_products', 'readingownedbook x reading_utilities', 'readingownedbook x downloads_per_book', 'readingownedbook x percent_reading', 'readingownedbook x percent_utilities', 'readingownedbook x events_per_book', 'ebookdownloaded x readingfreepreview', 'ebookdownloaded x highlightcreated', 'ebookdownloaded x freecontentcheckout', 'ebookdownloaded x readingopenchapter', 'ebookdownloaded x wishlistitemadded', 'ebookdownloaded x crossreferencetermopened', 'ebookdownloaded x total_events', 'ebookdownloaded x distinct_products', 'ebookdownloaded x reading_utilities', 'ebookdownloaded x downloads_per_book', 'ebookdownloaded x percent_reading', 'ebook

In [13]:
import statsmodels.api as sm
import numpy as np
import pandas as pd

all_features = features + interaction_features

# Define X and y
X = df[all_features]
y = df['Content Driven Customer']

# Add a constant to the model (Adding y-intercept)
X = sm.add_constant(X)

# Fit the OLS model
model_interaction = sm.OLS(y, X).fit()

print(model_interaction.summary())

# Summary table as a DataFrame

summary_df = pd.DataFrame({
    "Variable": model.params.index,
    "Coef": model.params.values,
    "Std Err": model.bse.values,
    "t": model.tvalues,
    "P>|t|": model.pvalues,
    "CI 2.5%": model.conf_int()[0],
    "CI 97.5%": model.conf_int()[1]
})

summary_df = model_interaction.summary2().tables[1]
print(summary_df)

                               OLS Regression Results                              
Dep. Variable:     Content Driven Customer   R-squared:                       0.399
Model:                                 OLS   Adj. R-squared:                  0.395
Method:                      Least Squares   F-statistic:                     118.8
Date:                     Thu, 13 Feb 2025   Prob (F-statistic):               0.00
Time:                             06:27:18   Log-Likelihood:                -9344.2
No. Observations:                    20000   AIC:                         1.891e+04
Df Residuals:                        19888   BIC:                         1.980e+04
Df Model:                              111                                         
Covariance Type:                 nonrobust                                         
                                                     coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------

# Task B: Questions and answers

**Changes in the R-squared (or other evaluation metrics).**

- Model without interaction variable : 0.305
- Model with interaction variable : 0.399


**Changes in the significance and magnitude of the main effects when interaction terms are introduced.**

 **1. Percent_reading** (the most predictive features in the original model)

Original Model:
- Coefficient: 0.6803
- P-value: 0.0000

Interaction Model:
- Coefficient: 0.4200
- P-value: 0.0000

Summary: Although the magnitude of percent_reading’s main effect decreases, it remains a strong and significant predictor, suggesting that its impact partly depends on the context provided by other variables.

**2. Percent_utilities**

Original Model:
- Coefficient: 0.5469
- P-value: 0.0000

Interaction Model:
- Coefficient: 0.3373
- P-value: 0.0000

Summary: The decrease in the coefficient implies that the role of percent_utilities is partly moderated by other features in the model.

**3. Distinct_products**

Original Model:
- Coefficient: 0.0232
- P-value: 0.0000

Interaction Model:
- Coefficient: 0.0371
- P-value: 0.0000

Summary: With interaction terms, the coefficient for distinct_products rises to 0.0371, indicating a stronger influence, likely due to its amplified effect when combined with other behavioral metrics.

> **Other remaining changes in the significance and magnitude of the main effects features:**

4) downloads_per_book:

Original Model:
- Coefficient: 0.0912
- P-value: 0.0000

Interaction:
- Coefficient 0.0633
- P-value: 0.0000

Summary: With interaction terms, the coefficient for downloads_per_book  decreases to 0.0633 is partly affected by other metrics in the model.


5) events_per_book:

Original Model:
- Coefficient: 0.0049
- P-value: 0.0000

Interaction:
- Coefficient: 0.0323
- P-value: 0.0000

Summary: The decrease in the coefficient implies that the role of events_per_book is partly impacted by other features in the model.






# **Q2. Which interactions turned out to be strong predictors?**

# A. Strong interaction predictors (p < 0.001 with high t-values) include:
- readingownedbook × ebookdownloaded: Coefficient = 0.0003, t ≈ 8.97
- readingownedbook × events_per_book: Coefficient = –6.481e-05, t ≈ –11.72
- ebookdownloaded × percent_reading: Coefficient = –0.0582, t ≈ –8.91
- downloads_per_book × percent_reading: Coefficient = 0.3288, t ≈ 15.89
- percent_reading × percent_utilities: Coefficient = 1.4523, t ≈ 11.92


These interactions indicate that the effect of one predictor significantly depends on another, enhancing the overall model’s explanatory power.

- readingownedbook × ebookdownloaded: Users who own physical books are also interested in eBooks. Rather than fully replacing them, they increased the engagement through digitalisation.

- readingownedbook × events_per_book: This shows an inverse relationship. Owning more books decreases the interest to engage in event because customers already have their preference on their own collection.

- ebookdownloaded × percent_reading: The negative coefficient indicates that ebooks tend to affect the reading percentage negatively. It does not result in increased reading activity.

- downloads_per_book × percent_reading: The strong positive relationship indicates that as number of download per book increases, the percent reading also increases. It indicates a high engagement in reading behavior among users.  

- percent_reading × percent_utilities: This shows that customers who spend more time on reading, use the reading utilities effictively. The high coefficient suggests that reading engagement relates with the digital tool usage and making this as the strongest key predictor in this model.

# Task C: Interpretation and Recommendations

# Q1. **Which Features Matter Most? Summarize the regression results**
- Positive drivers (can be individual features or interaction of features) of a user being content-driven.
> - "Percent_reading" : +0.4200 (t = 22.06, p < 0.001)
> - "Distinct_products" : +0.0371 (t = 14.42, p < 0.001)
> - "Downloads_per_book" : +0.0633 (t = 10.05, p < 0.001)


- Negative drivers (features that indicate price sensitivity).
> - "Reading Owned Books × Downloads per Book" : –0.0068 (t = –6.34, p < 0.001)
> - "Reading Owned Books × Events per Book" : –6.48e-05 (t = –11.72, p < 0.001)
> - "eBook Downloaded × Percent Reading" : –0.0582 (t = –8.91, p < 0.001)

The model sorts customers into two groups: content-driven (1) or price-driven (0), based on their survey answers. If a customer is highly engaged with content, they get a 1, meaning those with lower engagement are seen as price-driven.

However, the negative drivers in the model simply show less content engagement—they do not directly prove price sensitivity. So, assuming that "not content-driven" automatically means "price-driven" is just a way to simplify things, but real customer behavior may be more complicated.

# Q2. **What Are the Practical Implications? Suppose you’re presenting to the marketing team**
>- How might you **segment customers** or **tailor promotions** based on these findings?  

Based on the findings and regression outcome, customers can be segmented into content-driven and prive-driven readers. Both groups of readers can be approached with a well-planned marketing strategies.

Content-driven readers:
- Segment: Engaged readers
- Features: High percent_reading, high distinct_products, high downloads_per_book
- Marketing Strategy: To offer exclusive early access with exclusive contents to attract loyalty and eagerness to be a part of the reading community based one each preferences.

Price-driven readers:
- Segment: Readers who seek for bargains
- Features: Low percent_reading, low events_per_book, high reading_books
- Marketing Strategy: To provide discounts and promotions within limited-time with exclusive highlights or contents.


>- Which features or behaviors would you track more closely going forward?
As for marketing strategies, our team would:

1. Track and analyse the engagement metrics
- As proven from the regression output above, we would monitor the percent_reading with percent_utilties. Example, if users percent_reading is high and they use utilities such as annotations, highlights or bookmarks, it defines their reading behavior after promotional events.

2. To understand customers' download behavior
- Analysing the engagement metrics, we would identify their group categories to content-driven readers or price-driven readers. For example, to analyse whether customers fully utilise the percent_reading based on the downloads_per_book.

For more detailed explanations based on tasks and questions given, it is answered on the report attached in the assignment submission.